### PR TITLE
bump avro to 1.11.1

### PR DESCRIPTION
**Pull request by ngilles**
# About this change - What it does

* enum default value must be an element of enum symbols
* Invalid binary representations trigger a different exception
* removed test for "reserved" names

References: #439 

# Why this way

### CODE REVIEW

**Review by tvainika — CHANGES_REQUESTED** [review]
Nice. Due some specific behaviours in Avro there has been some special handling in Karapace side. Some of your proposed changes to tests need a bit polishing. Also after Avro 1.11.1 there has been upstream submitted and merged namespace related fix to Python side. That also causes more work on Karapace side to avoid regression with old data written with earlier Karapace releases.

File: tests/integration/test_schema.py [hunk 1]
```diff
@@ -2828,11 +2828,3 @@ async def test_invalid_schema_should_provide_good_error_messages(registry_async_
...
         == "Invalid AVRO schema. Error: Enum symbols must be a sequence of strings, but it is <class 'dict'>"
     )
-
-    # This is an upstream bug in the python AVRO library, until the bug is fixed we should at least have a nice error message
-    schema_str = json.dumps({"type": "enum", "name": "error", "symbols": ["A", "B"]})
-    res = await registry_async_client.post(
-        f"subjects/{test_subject}/versions",
-        json={"schema": schema_str},
-    )
-    assert res.json()["message"] == "Invalid AVRO schema. Error: error is a reserved type name."
```
[the remark this answers is not in the document]
Comment (ngilles) [reply tests/integration/test_schema.py:2838]: Yes, as far as I understood there were some name that were "reserved" before (such as `error` and `record`, etc) which is not the case. I'll restore the test and "invert it"

File: tests/unit/test_avro_compatibility.py [hunk 2]
```diff
@@ -36,11 +36,11 @@
     '{"type":"record","name":"myrecord","fields":[{"type":["null","string"],"name":"f1","default":'
     '"null"},{"type":"string","name":"f2","default":"foo"},{"type":"string","name":"f3","default":"bar"}]}'
 )
-invalidEnumDefaultValue = parse_avro_schema_definition(
-    '{"type": "enum", "name": "test_default", "symbols": ["A"], "default": "B"}'
+incorrectEnumDefaultValue = parse_avro_schema_definition(
+    '{"type": "enum", "name": "test_default", "symbols": ["A", "B"], "default": "B"}'
```
[the remark this answers is not in the document]
Comment (ngilles) [reply tests/unit/test_avro_compatibility.py:40]: The original schema is now *actually* invalid (as in won't even parse as a valid schema).

The only place it is then used is to test if if `incorrectEnumDefaultValue` and `correctEnumDefaultValue` schemas are compatible, the test mentions just "changing enum default value should be comptible", which is what this change tries to maintain.

But there totally should be some other test about the newly invalid schemas (with enum default value not being part of enum symbols). I'm not 100% sure how to go about this one. Probably the test should be deeper (like, what if a previously provided schema now finds itself in an upgrade karapace)?

File: tests/unit/test_avro_compatibility.py [hunk 3]
```diff
@@ -36,11 +36,11 @@
...
 )
 correctEnumDefaultValue = parse_avro_schema_definition(
-    '{"type": "enum", "name": "test_default", "symbols": ["A"], "default": "A"}'
+    '{"type": "enum", "name": "test_default", "symbols": ["A", "B"], "default": "A"}'
```
[the remark this answers is not in the document]
Comment (ngilles) [reply tests/unit/test_avro_compatibility.py:43]: As above, the "test objective" as desrcibed in in the test case is:
```
msg = "changing a default value in enum is a backward compatible change"
```
Considering that by "backward compatible" we mean "A new schema is backward compatible if it can be used to read the data written in a all previous schemas", I thought the new schema would need to accept "the same or more values" than previous schema. So if we had to define "B" as valid symbol for `incorrectEnumDefaultValue` I assumed this was also needed.

I just tested by not adding it, and the test passes, which I think exposes a bug: If old schema allow `A | B`, and new one only allows `A` ... any data containing `B` would be invalid

File: tests/unit/test_avro_compatibility.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjaakola-aiven) [reply tests/unit/test_avro_compatibility.py:40]: The `ReaderWriterCompatibilityChecker.get_compatibility` requires a `Schema` object as argument. I think the original test in this file does not serve the original intent anymore.

The concern @tvainika has is valid and @ngilles your suggestion is the way to go. Introduce a test where an invalid schema is read by from `_schemas` topic and new schema posted where compatibility is checked against the invalid one.

File: tests/unit/test_avro_compatibility.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjaakola-aiven) [reply tests/unit/test_avro_compatibility.py:43]: The `A | B` to `A` sounds like an Avro defect.

**Review by jjaakola-aiven — CHANGES_REQUESTED** [review]

File: tests/unit/test_avro_compatibility.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RyanSkraba) [reply tests/unit/test_avro_compatibility.py:40]: > If new avro library rejects such state in some new way, it needs to be handled before merging this.

OK, it looks like this was tightened up in Avro 1.11.1.  The schema *as is* is definitely invalid if you were try to use it in evolution.  Nothing would prevent it from being used outside of evolution.

The Java SDK permits you to ignore defaults (in records) for example.  Having this escape route is important for many compatibility use cases.  I raised AVRO-3637 for this (much like you can currently ignore invalid enum names if you want in Python).

File: tests/unit/test_avro_compatibility.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RyanSkraba) [reply tests/unit/test_avro_compatibility.py:43]: Hello!  I checked into this and the spec is _really_ confusing :/  I had to go back to the PR discussion to reason through what it's saying.

*TL;DR(1)*: Deleting an enum symbol is a permitted, but non-backwards compatible evolution.  If you never encounter `B`, everything's fine.

*TL;DR(2)*: Deleting an enum symbol with a default IS a backwards compatible evolution.  If you encounter `B`, the default is used.

**Comment by kiliangrashoff** [comment]
Hi! @ngilles are you still working on this, if not could I help out to get this PR finished?

**Comment by ngilles** [comment]
@kiliangrashoff Unfortunately I haven't had much time to dedicate to this, and have been stalling, you can go ahead and adopt this one!

**Comment by tvainika** [comment]
Thank you for this work, your contribution was indirectly used now when Avro requirement was bumped to 1.11 branch + our own patch submitted upstream to be included in 1.11.2.
